### PR TITLE
[wip] feat(salesforce): Added salesforce connector

### LINKED ISSUE #1057: Salesforce Connector
__Is your feature request related to a problem? Please describe.__

I would like to use Camunda Platform to retrieve data from the Salesforce REST API.

__Describe the solution you'd like__

A Salesforce sObjects connector

__Describe alternatives you've considered__

Using the plain REST connector which is some redundancy as Salesforce offers a very structured approach to access the REST API

__Additional context__

@ManuelDittmar and I did a hackdays project and developed the connector there.

### PR DESCRIPTION
**Pull request by jonathanlukas**
## Description

I created a new connector which is abstracting the salesforce REST API.

This connector was used for our Summer Hackdays (with slight modifications).

## Related issues

closes #1057 

Docs: https://github.com/camunda/camunda-platform-docs/pull/2520

### CODE REVIEW

**Comment by jonathanlukas** [comment]
I would like to carry out some more testing on this, so please do not yet merge

File: connectors/salesforce/element-templates/salesforce-connector.json [hunk 1]
```diff
@@ -0,0 +1,434 @@
...
+        "notEmpty": true,
+        "pattern": {
+          "value": "^(=|http://|https://|secrets).*$",
+          "message": "Must be a http(s) URL."
+        }
+      }
+    },
+    {
+      "id": "salesforceOperationType",
+      "label": "Salesforce Operation Type",
```
Comment (igpetrov) [inline connectors/salesforce/element-templates/salesforce-connector.json:94]: Should follow the same style as other connectors:

```suggestion
      "label": "Salesforce operation type",
```

File: connectors/salesforce/element-templates/salesforce-connector.json [hunk 2]
```diff
@@ -0,0 +1,434 @@
...
+      "type": "String",
+      "feel": "optional",
+      "binding": {
+        "type": "zeebe:input",
+        "name": "baseUrl"
+      },
+      "constraints": {
+        "notEmpty": true,
+        "pattern": {
+          "value": "^(=|http://|https://|secrets).*$",
```
Comment (igpetrov) [inline connectors/salesforce/element-templates/salesforce-connector.json:87]: @Oleksiivanov could you please take a look and say whether this pattern still makes sense?

File: connectors/salesforce/element-templates/salesforce-connector.json [hunk 3]
```diff
@@ -0,0 +1,434 @@
...
+          "value": "sObject"
+        },
+        {
+          "name": "SOQL Query",
+          "value": "soqlQuery"
+        }
+      ]
+    },
+    {
+      "label": "Interaction Type",
```
Comment (igpetrov) [inline connectors/salesforce/element-templates/salesforce-connector.json:114]:
```suggestion
      "label": "Interaction type",
```

File: connectors/salesforce/element-templates/salesforce-connector.json [hunk 4]
```diff
@@ -0,0 +1,434 @@
...
+      "id": "interactionType",
+      "group": "endpoint",
+      "description": "Choose the kind of interaction you want to perform",
+      "type": "Dropdown",
+      "choices": [
+        {
+          "name": "Get Object",
```
Comment (igpetrov) [inline connectors/salesforce/element-templates/salesforce-connector.json:121]:
```suggestion
          "name": "Get object",
```

File: connectors/salesforce/element-templates/salesforce-connector.json [hunk 5]
```diff
@@ -0,0 +1,434 @@
...
+          "value": "get"
+        },
+        {
+          "name": "Create Object",
```
Comment (igpetrov) [inline connectors/salesforce/element-templates/salesforce-connector.json:125]:
```suggestion
          "name": "Create object",
```

File: connectors/salesforce/element-templates/salesforce-connector.json [hunk 6]
```diff
@@ -0,0 +1,434 @@
...
+          "value": "post"
+        },
+        {
+          "name": "Update Object",
```
Comment (igpetrov) [inline connectors/salesforce/element-templates/salesforce-connector.json:129]:
```suggestion
          "name": "Update object",
```

File: connectors/salesforce/element-templates/salesforce-connector.json [hunk 7]
```diff
@@ -0,0 +1,434 @@
...
+          "value": "patch"
+        },
+        {
+          "name": "Delete Object",
```
Comment (igpetrov) [inline connectors/salesforce/element-templates/salesforce-connector.json:133]: Should follow the same style as other connectors:

```suggestion
          "name": "Delete object",
```

File: connectors/salesforce/element-templates/salesforce-connector.json [hunk 8]
```diff
@@ -0,0 +1,434 @@
...
+        "type": "zeebe:input",
+        "name": "method"
+      },
+      "condition": {
+        "equals": "soqlQuery",
+        "property": "salesforceOperationType"
+      }
+    },
+    {
+      "label": "Salesforce Object Type",
```
Comment (igpetrov) [inline connectors/salesforce/element-templates/salesforce-connector.json:162]:
```suggestion
      "label": "Salesforce object type",
```

File: connectors/salesforce/element-templates/salesforce-connector.json [hunk 9]
```diff
@@ -0,0 +1,434 @@
...
+      "description": "Type of the Salesforce Object",
```
Comment (igpetrov) [inline connectors/salesforce/element-templates/salesforce-connector.json:163]:
```suggestion
      "description": "Type of the Salesforce object",
```

File: connectors/salesforce/element-templates/salesforce-connector.json [hunk 10]
```diff
@@ -0,0 +1,434 @@
...
+      "constraints": {
+        "notEmpty": true
+      },
+      "condition": {
+        "equals": "sObject",
+        "property": "salesforceOperationType"
+      }
+    },
+    {
+      "label": "Salesforce Object Id",
```
Comment (igpetrov) [inline connectors/salesforce/element-templates/salesforce-connector.json:180]:
```suggestion
      "label": "Salesforce object ID",
```

File: connectors/salesforce/element-templates/salesforce-connector.json [hunk 11]
```diff
@@ -0,0 +1,434 @@
...
+            "equals": "sObject"
+          }
+        ]
+      },
+      "constraints": {
+        "notEmpty": true
+      }
+    },
+    {
+      "label": "Relation Name",
```
Comment (igpetrov) [inline connectors/salesforce/element-templates/salesforce-connector.json:210]:
```suggestion
      "label": "Relation name",
```

File: connectors/salesforce/element-templates/salesforce-connector.json [hunk 12]
```diff
@@ -0,0 +1,434 @@
...
+        "name": "url"
+      },
+      "value": "=baseUrl + \"/services/data/v58.0/query\"",
+      "condition": {
+        "equals": "soqlQuery",
+        "property": "salesforceOperationType"
+      }
+    },
+    {
+      "label": "Request Body",
```
Comment (igpetrov) [inline connectors/salesforce/element-templates/salesforce-connector.json:279]:
```suggestion
      "label": "Request body",
```

File: connectors/salesforce/element-templates/salesforce-connector.json
[the remark this answers is not in the document]
Comment (jonathanlukas) [reply connectors/salesforce/element-templates/salesforce-connector.json:1]: Was added

File: connectors/salesforce/element-templates/salesforce-connector.json [hunk 13]
```diff
@@ -0,0 +1,434 @@
+{
+  "$schema": "https://unpkg.com/@camunda/zeebe-element-templates-json-schema/resources/schema.json",
+  "name": "Salesforce sObject Connector",
+  "id": "90f7ab88-3924-45ed-8f04-66441748f35f",
+  "description": "Interact with Salesforce sObjects",
```
[the remark this answers is not in the document]
Comment (jonathanlukas) [reply connectors/salesforce/element-templates/salesforce-connector.json]: The description is now more generic, as the connector does not only handle sobjects

File: connectors/salesforce/element-templates/salesforce-connector.json [hunk 14]
```diff
@@ -0,0 +1,434 @@
...
...CAtMiAyNzcgNDAgODAgMTQ1IDEwNiAyMzEgNTh6Ii8+CjxwYXRoIGQ9Ik01ODE4IDM2MDYgYy0yMSAtOCAtNDkgLTIxIC02MyAtMzAgLTMyIC0yMSAtMzUgLTIwIC0zNSA5IDAgMjYgLTEKMjYgLTY3IDIzIGwtNjggLTMgMCAtMzI1IDAgLTMyNSA3MCAwIDcwIDAgNSAyMTUgYzQgMTgyIDggMjIwIDIzIDI0OCAyNyA1MAo3MSA3MiAxNDMgNzIgbDYxIDAgMTcgNDMgYzI1IDY2IDIyIDczIC0zMCA4MSAtNjQgOCAtODMgNyAtMTI2IC04eiIvPgo8cGF0aCBkPSJNNjI1MSAzNjA0IGMtMTQzIC0zOCAtMjIwIC0xNDAgLTIyOSAtMzAxIC0xNCAtMjY2IDE4OCAtNDA0IDQ4NgotMzMxIDQ3IDEyIDQ5IDE5IDI2IDc4IC0xMyAzMSAtMjEgMzkgLTMzIDM0IC0zMiAtMTMgLTE2OCAtMTYgLTIwNCAtNSAtODcgMjkKLTEyNyA5NCAtMTI3IDIwNiAwIDgzIDIyIDEzOCA2OSAxNzggNDQgMzYgNjcgNDEgMTgwIDM0IGw5NyAtNiAxNyA0NiBjMjEgNTUKMTkgNTcgLTY4IDcyIC04MiAxNCAtMTQ4IDEzIC0yMTQgLTV6Ii8+CjxwYXRoIGQ9Ik02ODIyIDM2MDUgYy0xNTUgLTQzIC0yMzEgLTE2NiAtMjIwIC0zNTQgOCAtMTQ0IDc4IC0yMzkgMjA2IC0yNzgKMTExIC0zNCAzNzIgLTEyIDM3MiAzMiAwIDEyIC0yOSA5NSAtMzQgOTUgMCAwIC0yMyAtNyAtNTAgLTE1IC0yNyAtOCAtODYgLTE1Ci0xMzAgLTE1IC0xMzAgMCAtMTkzIDQzIC0yMTEgMTQyIGwtNyAzOCAyMjggMiAyMjkgMyAtMSA4MCBjLTEgMTI2IC01MiAyMTIKLTE1MiAyNTcgLTU3IDI2IC0xNjMgMzIgLTIzMCAxM3ogbTE1OCAtMTExIGM0NCAtMTggNzcgLTYxIDgyIC0xMDkgbDMgLTMwCi0xNDUgLTMgYy03OSAtMSAtMTUwIDAgLTE1NyAyIC05IDQgLTEwIDEyIC0xIDM4IDM1IDEwMyAxMTggMTQyIDIxOCAxMDJ6Ii8+CjwvZz4KPC9zdmc+Cg=="
```
[the remark this answers is not in the document]
Comment (jonathanlukas) [reply connectors/salesforce/element-templates/salesforce-connector.json]: the svg is now 18x18

File: connectors/salesforce/element-templates/salesforce-connector.json [hunk 15]
```diff
@@ -0,0 +1,434 @@
+{
+  "$schema": "https://unpkg.com/@camunda/zeebe-element-templates-json-schema/resources/schema.json",
+  "name": "Salesforce sObject Connector",
+  "id": "90f7ab88-3924-45ed-8f04-66441748f35f",
```
[the remark this answers is not in the document]
Comment (jonathanlukas) [reply connectors/salesforce/element-templates/salesforce-connector.json]: The id is now more specific

**Comment by sbuettner** [comment]
Hi @jonathanlukas, first of all thanks for the contribution and the effort to work on this protocol connector. Thats exactly what we like to see as this is the most sustainable way to create a new connector.

After discussing the topic of adding more element templates to this repository we came to the conclusion that we dont want to add more element templates to this repository as this would mean to add it to our list of officially maintained connectors.

Therefore we would be really happy if you could move this connector to the community hub at https://github.com/orgs/camunda-community-hub/repositories?q=connector&type=all&language=&sort= and add it to the list of connectors at: https://github.com/camunda-community-hub/camunda-8-connectors

Thank you! 🤖

File: connectors/salesforce/element-templates/salesforce-connector.json (hunk 2 above)
Comment (jonathanlukas) [reply connectors/salesforce/element-templates/salesforce-connector.json:88]: Ok. Will do. Did you already adjust the pattern in the http connector? I started by copying the http connector and did not change the pattern

**Comment by sbuettner** [comment]
Hey @jonathanlukas, its not about whether we think its not important. Its more about who will be responsible for managing feature requests and does maintenance work in case of API changes and new requirements. Thats why we are currently thinking how and where we manage protocol connectors in general as this repository will mainly be used for pro-code connectors.

**Comment by sbuettner** [comment]
@jonathanlukas All good. I just want to make sure that you dont get impression we dont want it. We are supper happy about any protocol connector contribution. 

@bastiankoerber Should we rather create a connectors template mono-repo for all the new and upcoming connectors to avoid too many repos?

**Comment by sbuettner** [comment]
@jonathanlukas I discussed this topic again with @bastiankoerber and we came to the conclusion that the connector had quite some requests already and we would therefore merge it into the official connectors repo.

Thanks for your contribution.

@igpetrov Please take a final look and we can merge this one.

**Comment by igpetrov** [comment]
Taking a look

**Comment by igpetrov** [comment]
Hi @jonathanlukas 

LGTM. As per [comment](https://github.com/camunda/connectors/pull/1059/files#r1299971145), all official Connectors should have their own doc page. Could you please create it and update the link?
Alternatively, you can update the link now, and write the doc later but it should before before 12th of September, since that's the release date.

The doc shouldn't be extensive, but overal explaining how to use the connector. You can use [Asana connector doc](https://github.com/camunda/camunda-platform-docs/commit/35584406b09dc098b1d80a242c13bdc55c064ee9) as reference.

Should you have further questions, please reach out to me.

Thank you.

File: connectors/salesforce/element-templates/salesforce-connector.json [hunk 16]
```diff
@@ -0,0 +1,434 @@
...
+  },
+  "documentationRef": "https://docs.camunda.io/docs/components/connectors/out-of-the-box-connectors/rest/",
```
[the remark this answers is not in the document]
Comment (jonathanlukas) [reply connectors/salesforce/element-templates/salesforce-connector.json]: Has been adjusted

**Review by igpetrov — APPROVED** [review]
LGTM, thank you for you contribution!

**Pull request merged**